### DISCUSSION TITLE
Pass the data in a similar way to validation rules when a resource is used via relationship manager

### DISCUSSION THREAD

**Discussion by elcapo** [filamentphp/filament#6521] [Ideas]
# Scenario

You might want to use a validation rule in order to implement a validation that depends on more than one field. For instance, you when validating the zip code of an address than also has a country field, the validation of the zip code depends on the country that was selected.

## The Issue

As the examples below are showing, the ways to access data from a validation rule are different depending on where the validation is being used from.

## The Question

Can the validation rules be used by **Filament** so that the `data` is always passed to validation rules in the same way?

The examples show that a validation that involves a column named **column_name** will be different depending on where it's being used from:

~~~php
/* If the rule is being used from a standard Laravel request validation, etc */
$data['column_name'];

/* If the rule is being used from a standard Filament resource */
$data['data']['column_name'];

/* If the rule is being used from a Filament resource that's being managed via relationship */
$data['mountedTableActionData']['column_name'];
~~~

## Example Rule

To provide a real world example, I'm using below the [ronanguilloux/IsoCodes](https://github.com/ronanguilloux/IsoCodes) library to validate a zip code.

~~~php
<?php

namespace App\Rules;

use App\Models\Country;
use Illuminate\Contracts\Validation\DataAwareRule;
use Illuminate\Contracts\Validation\ValidationRule;
use IsoCodes\ZipCode;

class ZipCodeRule implements ValidationRule, DataAwareRule
{
    protected ?Country $country = null;
    protected ?string $zipCode = null;

    /**
     * Run the validation rule.
     *
     * @param  \Closure(string): \Illuminate\Translation\PotentiallyTranslatedString  $fail
     */
    public function validate(string $attribute, mixed $value, \Closure $fail): void
    {
        if (!$this->zipCode || !$this->country->iso_code) {
            return;
        }

        if (!ZipCode::validate($this->zipCode, $this->country->iso_code)) {
            $fail(__("The zip code format does not correspond with {$this->country->name}'s rules."));
        }
    }

    /**
     * Set the data under validation.
     *
     * @param  array<string, mixed>  $data
     */
    public function setData(array $data): static
    {
        /* The issue comes with this function, so I'll provide two examples below */
    }
}
~~~

### Example Rule in a Standard Resource

To be able to use the rule in a resource like this:

~~~php
Forms\Components\TextInput::make('zip_code')
    ->rules([new App\Rules\ZipCodeRule()]),
~~~

The **getData** method had to access the data as in `$data['data']`:

~~~php
public function setData(array $data): static
{
    $countryId = $data['data']['country_id'];

    if (!$countryId) {
        return $this;
    }

    $this->country = Country::whereId($countryId)->first();
    $this->zipCode = $data['data']['zip_code'];

    return $this;
}
~~~

### Example Rule in a Managed Relationship

But the weird part is that, to be able to use the rule in a resource that's being managed via relationship, the data has to be accessed as `$data['mountedTableActionData']` :

~~~php
public function setData(array $data): static
{
    $countryId = $data['mountedTableActionData']['country_id'];

    if (!$countryId) {
        return $this;
    }

    $this->country = Country::whereId($countryId)->first();
    $this->zipCode = $data['mountedTableActionData']['zip_code'];

    return $this;
}
~~~

**Comment by danharrin** [comment]
You don't actually need anything like that

Filament has a set of utilities that are injected everywhere. https://filamentphp.com/docs/2.x/forms/advanced#using-closure-customization

One of them is `$get()`. You do not need to worry about `$data`, the fields will appear in different places inconsistently. `$get` is relative to the current field.

```php
->rule(fn ($get) => new MyCustomRule($get('zip_code')))
```

**Reply by elcapo** [reply]
Thanks @danharrin!

I didn't understand this from the documentation but now that I get it, the code looks much clearer.

Just for the record (there may be other people who didn't get it at first, like me), the class does not need to be a **DataAwareRule** as it would receive the information on construction.

~~~php
class ZipCodeRule implements ValidationRule
{
    protected ?Country $country;
    protected ?string $zipCode;

    public function __construct(string $zipCode, string $countryId)
    {
        $this->country = Country::whereId($countryId)->first();
        $this->zipCode = $zipCode;
    }

    /* No need to add a getData method */
    /* No changes in the validate method */
}
~~~
